MueLu:  Build error in MueLu_BelosSmoother

## Bug Report
@trilinos/muelu

### Description
We're seeing a build error in all of our integration builds (gcc, intel, and nvidia).

`Trilinos/packages/muelu/src/Smoothers/MueLu_BelosSmoother_decl.hpp:72:10: fatal error: BelosXpetraAdapter.hpp: No such file or directory`

### Steps to Reproduce
Configure script:
```
cmake \
  -G "Ninja" \
  -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON \
  -D Trilinos_EXTRA_LINK_FLAGS:STRING="" \
  -D CMAKE_BUILD_TYPE:STRING=release \
 -D Trilinos_ENABLE_OpenMP:BOOL=OFF \
 -D TPL_ENABLE_CUDA:BOOL=OFF \
 -DCMAKE_C_FLAGS_RELEASE_OVERRIDE:STRING="-DNDEBUG" \
 -DCMAKE_CXX_FLAGS_RELEASE_OVERRIDE:STRING="-DNDEBUG" \
 -DCMAKE_Fortran_FLAGS_RELEASE_OVERRIDE:STRING="-DNDEBUG" \
 -DCMAKE_C_FLAGS:STRING="-O2 -finline-functions -ftemplate-depth-128 -w " \
 -DCMAKE_CXX_FLAGS:STRING="-O2 -finline-functions -ftemplate-depth-128 -w " \
 -DCMAKE_Fortran_FLAGS:STRING="-O2 -finline-functions -ftemplate-depth-128 -w  " \
 -DHAVE_GCC_ABI_DEMANGLE:BOOL=ON \
 -DTrilinos_WARNINGS_AS_ERRORS_FLAGS:STRING="" \
 -DTrilinos_SHOW_DEPRECATED_WARNINGS:BOOL=OFF \
 -DTrilinos_ASSERT_MISSING_PACKAGES:BOOL=OFF \
 -DTrilinos_ENABLE_COMPLEX_DOUBLE:BOOL=ON \
 -DTrilinos_ENABLE_FLOAT:BOOL=ON \
 -DTrilinos_ENABLE_LINEAR_SOLVER_FACTORY_REGISTRATION:BOOL=ON \
 -DTrilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
 -DTrilinos_ENABLE_ALL_PACKAGES:BOOL=OFF \
 -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
 -DTrilinos_CXX11_FLAGS=-std=c++11 \
 -DTPL_ENABLE_Pthread:BOOL=OFF \
 -DTPL_ENABLE_yaml-cpp:BOOL=OFF \
 -DTPL_ENABLE_MPI:BOOL=ON \
 -DTPL_ENABLE_ParMETIS:BOOL=ON \
 -DTPL_ENABLE_UMFPACK:BOOL=ON \
 -DTPL_ENABLE_y12m:BOOL=ON \
 -DTPL_ENABLE_Boost:BOOL=ON \
 -DTrilinos_ENABLE_Amesos:BOOL=ON \
 -DTrilinos_ENABLE_Amesos2:BOOL=ON \
 -DTrilinos_ENABLE_Anasazi:BOOL=ON \
 -DTrilinos_ENABLE_AztecOO:BOOL=ON \
 -DTrilinos_ENABLE_Belos:BOOL=ON \
 -DTrilinos_ENABLE_Epetra:BOOL=ON \
 -DTrilinos_ENABLE_EpetraExt:BOOL=ON \
 -DTrilinos_ENABLE_FEI:BOOL=ON \
 -DTrilinos_ENABLE_Ifpack:BOOL=ON \
 -DTrilinos_ENABLE_Ifpack2:BOOL=ON \
 -DTrilinos_ENABLE_Intrepid:BOOL=ON \
 -DTrilinos_ENABLE_Kokkos:BOOL=ON \
 -DTrilinos_ENABLE_KokkosCore:BOOL=ON \
 -DTrilinos_ENABLE_KokkosContainers:BOOL=ON \
 -DTrilinos_ENABLE_MiniTensor:BOOL=ON \
 -DTrilinos_ENABLE_ML:BOOL=ON \
 -DTrilinos_ENABLE_MueLu:BOOL=ON \
 -DTrilinos_ENABLE_Pamgen:BOOL=ON \
 -DTrilinos_ENABLE_ROL:BOOL=ON \
 -DTrilinos_ENABLE_RTOp:BOOL=ON \
 -DTrilinos_ENABLE_Sacado:BOOL=ON \
 -DTrilinos_ENABLE_Shards:BOOL=ON \
 -DTrilinos_ENABLE_ShyLU_NodeTacho:BOOL=ON \
 -DTrilinos_ENABLE_ShyLU_NodeHTS:BOOL=ON \
 -DTrilinos_ENABLE_Stokhos:BOOL=OFF \
 -DTrilinos_ENABLE_Teuchos:BOOL=ON \
 -DTrilinos_ENABLE_ThreadPool:BOOL=ON \
 -DTrilinos_ENABLE_Thyra:BOOL=ON \
 -DTrilinos_ENABLE_ThyraEpetraExtAdapters:BOOL=ON \
 -DTrilinos_ENABLE_Tpetra:BOOL=ON \
 -DTrilinos_ENABLE_Triutils:BOOL=ON \
 -DTrilinos_ENABLE_Xpetra:BOOL=ON \
 -DTrilinos_ENABLE_Zoltan:BOOL=ON \
 -DTrilinos_ENABLE_Zoltan2:BOOL=ON \
 -DTrilinos_ENABLE_Galeri:BOOL=OFF \
 -DTrilinos_ENABLE_Gtest:BOOL=OFF \
 -DTrilinos_ENABLE_Isorropia:BOOL=OFF \
 -DTrilinos_ENABLE_MOOCHO:BOOL=OFF \
 -DTrilinos_ENABLE_Rythmos:BOOL=OFF \
 -DTrilinos_ENABLE_SEACAS:BOOL=OFF \
 -DTrilinos_ENABLE_STK:BOOL=OFF \
 -DTrilinos_ENABLE_Stratimikos:BOOL=OFF \
 -DTrilinos_ENABLE_TrilinosCouplings:BOOL=OFF \
 -DAmesos_ENABLE_SuperLU:BOOL=ON \
 -DAmesos2_ENABLE_ShyLU_NodeTacho:BOOL=OFF \
 -DAmesos2_ENABLE_SuperLU:BOOL=ON \
 -DAztecOO_ENABLE_AZLU:BOOL=ON \
 -DBelos_ENABLE_Xpetra:BOOL=OFF \
 -DIfpack2_ENABLE_ThyraTpetraAdapters=OFF \
 -DKOKKOS_ENABLE_DEPRECATED_CODE:BOOL=OFF \
 -DKOKKOS_ENABLE_PROFILING:BOOL=ON \
 -DML_ENABLE_METIS:BOOL=OFF \
 -DML_ENABLE_ParMETIS:BOOL=OFF \
 -DPamgen_ENABLE_Boost:BOOL=OFF \
 -DPamgen_ENABLE_Extras:BOOL=OFF \
 -DROL_ENABLE_EXAMPLES:BOOL=OFF \
 -DROL_ENABLE_TESTS:BOOL=OFF \
 -DSacado_ENABLE_KokkosCore:BOOL=ON \
 -DTacho_ENABLE_INT_INT:BOOL=ON \
 -DTpetra_INST_INT_LONG:BOOL=ON \
 -DTpetra_INST_INT_LONG_LONG:BOOL=OFF \
 -DTpetra_INST_SERIAL:BOOL=ON \
 -DZoltan_ENABLE_METIS:BOOL=OFF \
 -DZoltan_ENABLE_Scotch:BOOL=OFF \
 -DZoltan2_ENABLE_Experimental:BOOL=ON \
 -DZoltan2_ENABLE_Scotch:BOOL=OFF
```

--- a/packages/muelu/adapters/CMakeLists.txt
+++ b/packages/muelu/adapters/CMakeLists.txt
@@ -20,12 +20,6 @@ ENDIF()
 #
 IF (${PACKAGE_NAME}_ENABLE_Belos)
 
-  ASSERT_DEFINED(Belos_ENABLE_Xpetra)
-
-  IF (NOT Belos_ENABLE_Xpetra)
-    MESSAGE(FATAL_ERROR "MueLu requires the Belos Xpetra adapters in order to work with Belos. Please set \"Belos_ENABLE_Xpetra:BOOL=ON\"")
-  ENDIF ()
-
   INCLUDE_DIRECTORIES(${CMAKE_CURRENT_SOURCE_DIR}/belos)
 
   APPEND_SET(HEADERS

--- a/packages/muelu/src/Smoothers/MueLu_BelosSmoother_decl.hpp
+++ b/packages/muelu/src/Smoothers/MueLu_BelosSmoother_decl.hpp
@@ -48,19 +48,20 @@
 
 #include <Teuchos_ParameterList.hpp>
 
-#include <Xpetra_BlockedCrsMatrix_fwd.hpp>
-#include <Xpetra_CrsMatrixWrap.hpp>
 #include <Xpetra_Matrix_fwd.hpp>
-#include <Xpetra_Matrix.hpp>
 #include <Xpetra_MultiVectorFactory_fwd.hpp>
-#ifdef HAVE_XPETRA_TPETRA
-#include <Xpetra_TpetraCrsMatrix.hpp>
-#endif
 
 #include "MueLu_ConfigDefs.hpp"
-#include "MueLu_BelosSmoother_fwd.hpp"
 
 #if defined(HAVE_MUELU_BELOS)
+
+#include "MueLu_BelosSmoother_fwd.hpp"
+
+#ifdef HAVE_XPETRA_TPETRA
+#include "BelosTpetraAdapter.hpp"
+#include <Tpetra_Operator.hpp>
+#include <Tpetra_MultiVector.hpp>
+#endif
 
 #include "MueLu_FactoryBase_fwd.hpp"
 #include "MueLu_FactoryManagerBase_fwd.hpp"
@@ -69,7 +70,6 @@
 #include "MueLu_Utilities_fwd.hpp"
 
 #include <BelosSolverManager.hpp>
-#include <BelosXpetraAdapter.hpp>
 
 
 namespace MueLu {
@@ -190,17 +190,56 @@ namespace MueLu {
     RCP<Belos::SolverManager<Scalar, tMV, tOP> > tSolver_;
 #endif
 
-#ifdef HAVE_MUELU_EPETRA
-    typedef Epetra_MultiVector eMV;
-    typedef Epetra_Operator    eOP;
-    RCP<Belos::LinearProblem<Scalar, eMV, eOP> > eBelosProblem_;
-    RCP<Belos::SolverManager<Scalar, eMV, eOP> > eSolver_;
-#endif
-
     //! matrix, used in apply if solving residual equation
     RCP<Matrix> A_;
 
   }; // class BelosSmoother
+
+
+  #ifdef HAVE_MUELU_EPETRA
+
+# if ((defined(EPETRA_HAVE_OMP) && (!defined(HAVE_TPETRA_INST_OPENMP) || !defined(HAVE_TPETRA_INST_INT_INT))) || \
+    (!defined(EPETRA_HAVE_OMP) && (!defined(HAVE_TPETRA_INST_SERIAL) || !defined(HAVE_TPETRA_INST_INT_INT))))
+  // Stub specialization for missing Epetra template args
+  template<>
+  class BelosSmoother<double,int,int,Xpetra::EpetraNode> : public SmootherPrototype<double,int,int,Xpetra::EpetraNode> {
+    typedef double              Scalar;
+    typedef int                 LocalOrdinal;
+    typedef int                 GlobalOrdinal;
+    typedef Xpetra::EpetraNode  Node;
+#undef MUELU_BELOSSMOOTHER_SHORT
+#include "MueLu_UseShortNames.hpp"
+
+  public:
+#ifndef _MSC_VER
+    // Avoid error C3772: invalid friend template declaration
+    template<class Scalar2, class LocalOrdinal2, class GlobalOrdinal2, class Node2>
+    friend class BelosSmoother;
+#endif
+
+    BelosSmoother(const std::string& type, const Teuchos::ParameterList& paramList = Teuchos::ParameterList(), const LocalOrdinal& overlap = 0) {
+      MUELU_TPETRA_ETI_EXCEPTION("BelosSmoother<double,int,int,EpetraNode>","BelosSmoother<double,int,int,EpetraNode>","int");
+    };
+
+    virtual ~BelosSmoother() { }
+
+    void SetParameterList(const Teuchos::ParameterList& paramList) {}
+    void DeclareInput(Level &currentLevel) const {}
+    void Setup(Level &currentLevel) {}
+    void Apply(MultiVector& X, const MultiVector& B, bool InitialGuessIsZero = false) const {}
+    RCP<SmootherPrototype> Copy() const { return Teuchos::null;}
+
+    std::string description() const { return std::string(""); }
+    void print(Teuchos::FancyOStream &out, const VerbLevel verbLevel = Default) const {}
+
+    //! Get a rough estimate of cost per iteration
+    size_t getNodeSmootherComplexity() const {size_t cplx=0; return cplx;};
+
+  };
+# endif
+
+#endif // HAVE_MUELU_EPETRA
+
 
 } // namespace MueLu
 

--- a/packages/muelu/src/Smoothers/MueLu_BelosSmoother_def.hpp
+++ b/packages/muelu/src/Smoothers/MueLu_BelosSmoother_def.hpp
@@ -52,14 +52,12 @@
 
 #include <Teuchos_ParameterList.hpp>
 
-#include <Tpetra_RowMatrix.hpp>
-
-#include <Xpetra_BlockedCrsMatrix.hpp>
 #include <Xpetra_CrsMatrix.hpp>
-#include <Xpetra_CrsMatrixWrap.hpp>
 #include <Xpetra_Matrix.hpp>
 #include <Xpetra_MultiVectorFactory.hpp>
+#ifdef HAVE_XPETRA_TPETRA
 #include <Xpetra_TpetraMultiVector.hpp>
+#endif
 
 #include "MueLu_BelosSmoother_decl.hpp"
 #include "MueLu_Level.hpp"
@@ -82,10 +80,6 @@ namespace MueLu {
 #ifdef HAVE_MUELU_TPETRA
     Belos::SolverFactory<Scalar,tMV,tOP> tFactory;
     solverSupported = solverSupported || tFactory.isSupported(type);
-#endif
-#ifdef HAVE_MUELU_EPETRA
-    Belos::SolverFactory<Scalar,eMV,eOP> eFactory;
-    solverSupported = solverSupported || eFactory.isSupported(type);
 #endif
     // if (!solverSupported) {
     //   Teuchos::Array<std::string> supportedSolverNames = factory.supportedSolverNames();
@@ -145,15 +139,7 @@ namespace MueLu {
       tSolver_->setProblem(tBelosProblem_);
 #endif
     } else {
-#ifdef HAVE_MUELU_EPETRA
-      eBelosProblem_ = rcp(new Belos::LinearProblem<Scalar, eMV, eOP>());
-      RCP<eOP> eA = Utilities::Op2NonConstEpetraCrs(A_);
-      eBelosProblem_->setOperator(eA);
-
-      Belos::SolverFactory<SC, eMV, eOP> solverFactory;
-      eSolver_ = solverFactory.create(type_, rcpFromRef(const_cast<ParameterList&>(this->GetParameterList())));
-      eSolver_->setProblem(eBelosProblem_);
-#endif
+      TEUCHOS_ASSERT(false);
     }
   }
 
@@ -189,31 +175,7 @@ namespace MueLu {
       }
 #endif
     } else {
-#ifdef HAVE_MUELU_EPETRA
-    if (InitialGuessIsZero) {
-      X.putScalar(0.0);
-
-      RCP<Epetra_MultiVector>       epX = rcpFromRef(Utilities::MV2NonConstEpetraMV(X));
-      RCP<const Epetra_MultiVector> epB = rcpFromRef(Utilities::MV2EpetraMV(B));
-
-      eBelosProblem_->setInitResVec(epB);
-      eBelosProblem_->setProblem(epX, epB);
-      eSolver_->solve();
-
-    } else {
-      RCP<MultiVector> Residual   = Utilities::Residual(*A_, X, B);
-      RCP<MultiVector> Correction = MultiVectorFactory::Build(A_->getDomainMap(), X.getNumVectors());
-
-      RCP<Epetra_MultiVector>       epX = rcpFromRef(Utilities::MV2NonConstEpetraMV(*Correction));
-      RCP<const Epetra_MultiVector> epB = rcpFromRef(Utilities::MV2EpetraMV(*Residual));
-
-      eBelosProblem_->setInitResVec(epB);
-      eBelosProblem_->setProblem(epX, epB);
-      eSolver_->solve();
-
-      X.update(1.0, *Correction, 1.0);
-    }
-#endif
+      TEUCHOS_ASSERT(false);
     }
 
   }
@@ -232,10 +194,6 @@ namespace MueLu {
       if (A_->getRowMap()->lib() == Xpetra::UseTpetra) {
 #ifdef MUELU_HAVE_TPETRA
         out << tSolver_->description();
-#endif
-      } else {
-#ifdef MUELU_HAVE_EPETRA
-        out << eSolver_->description();
 #endif
       }
     } else {
@@ -261,12 +219,6 @@ namespace MueLu {
         out << *tSolver_ << std::endl;
       }
 #endif
-#ifdef MUELU_HAVE_EPETRA
-      if (eSolver_ != Teuchos::null) {
-        Teuchos::OSTab tab2(out);
-        out << *eSolver_ << std::endl;
-      }
-#endif
     }
 
     if (verbLevel & Debug) {
@@ -275,12 +227,6 @@ namespace MueLu {
         out0 << "IsSetup: " << Teuchos::toString(SmootherPrototype::IsSetup()) << std::endl
              << "-" << std::endl
              << "RCP<solver_>: " << tSolver_ << std::endl;
-#endif
-      } else {
-#ifdef MUELU_HAVE_EPETRA
-        out0 << "IsSetup: " << Teuchos::toString(SmootherPrototype::IsSetup()) << std::endl
-             << "-" << std::endl
-             << "RCP<solver_>: " << eSolver_ << std::endl;
 #endif
       }
     }

--- a/packages/muelu/src/Smoothers/MueLu_BelosSmoother_def.hpp
+++ b/packages/muelu/src/Smoothers/MueLu_BelosSmoother_def.hpp
@@ -52,13 +52,14 @@
 
 #include <Teuchos_ParameterList.hpp>
 
+#include <Tpetra_RowMatrix.hpp>
+
+#include <Xpetra_BlockedCrsMatrix.hpp>
 #include <Xpetra_CrsMatrix.hpp>
 #include <Xpetra_CrsMatrixWrap.hpp>
 #include <Xpetra_Matrix.hpp>
 #include <Xpetra_MultiVectorFactory.hpp>
-#ifdef HAVE_XPETRA_TPETRA
 #include <Xpetra_TpetraMultiVector.hpp>
-#endif
 
 #include "MueLu_BelosSmoother_decl.hpp"
 #include "MueLu_Level.hpp"

--- a/packages/muelu/test/unit_tests/Smoothers/BelosSmoother.cpp
+++ b/packages/muelu/test/unit_tests/Smoothers/BelosSmoother.cpp
@@ -61,9 +61,12 @@ namespace MueLuTests {
 #   include <MueLu_UseShortNames.hpp>
     MUELU_TESTING_SET_OSTREAM;
     MUELU_TESTING_LIMIT_SCOPE(Scalar,GlobalOrdinal,Node);
+    MUELU_TEST_ONLY_FOR(Xpetra::UseTpetra) {
 
-    BelosSmoother smoother("Block CG", Teuchos::ParameterList());
-    testApplyNoSetup(smoother, out, success);
+      BelosSmoother smoother("Block CG", Teuchos::ParameterList());
+      testApplyNoSetup(smoother, out, success);
+
+    }
 
   }
 
@@ -73,18 +76,19 @@ namespace MueLuTests {
 #   include <MueLu_UseShortNames.hpp>
     MUELU_TESTING_SET_OSTREAM;
     MUELU_TESTING_LIMIT_SCOPE(Scalar,GlobalOrdinal,Node);
+    MUELU_TEST_ONLY_FOR(Xpetra::UseTpetra) {
 
-    Teuchos::ParameterList paramList;
-    paramList.set("Maximum Iterations", 10);
-    BelosSmoother smoother("Block CG", paramList);
+      Teuchos::ParameterList paramList;
+      paramList.set("Maximum Iterations", 10);
+      BelosSmoother smoother("Block CG", paramList);
 
 
-    typename Teuchos::ScalarTraits<SC>::magnitudeType residualNorms = testApply_A125_X1_RHS0(smoother, out, success);
+      typename Teuchos::ScalarTraits<SC>::magnitudeType residualNorms = testApply_A125_X1_RHS0(smoother, out, success);
 
-    RCP<const Teuchos::Comm<int> > comm = TestHelpers::Parameters::getDefaultComm();
-    const typename Teuchos::ScalarTraits<SC>::magnitudeType expectedNorm = 1.2856486930664495771e-01;
-    TEST_FLOATING_EQUALITY(residualNorms,expectedNorm,1e-12);
-
+      RCP<const Teuchos::Comm<int> > comm = TestHelpers::Parameters::getDefaultComm();
+      const typename Teuchos::ScalarTraits<SC>::magnitudeType expectedNorm = 1.2856486930664495771e-01;
+      TEST_FLOATING_EQUALITY(residualNorms,expectedNorm,1e-12);
+    }
 
   } // Block CG
 

